dvc/project.py

```diff
@@ -383,6 +384,17 @@ def run(
     ):
         from dvc.stage import Stage
 
+        if outs is None:
```

Comment:
yep, I hate all these ifs, some refactoring is required around this method. `outs = outs or []` has different semantics unfortunately - it will return the second `[]` in this case `[] or []` as far as I remember.

Reply:
It isn't the intended, `only if outs is empty, initialize its value`?

```python
outs = None                                                                   
outs = outs or []                                                             
# >>> outs
# []

outs = outs or ['x', 'y']
# >>> outs
# ['x', 'y']

outs = outs or []
# >>> outs
# ['x', 'y']
```

The only difference is that the `val = val or new_val` checks for "falsy" values instead of just `None` (so `outs = False or "x"` will assign `"x"`, even if `False` is a valid value).

Reply:
My intuition to is to keep it as is - don't optimize by changing semantics just because it looks better. It breaks the protocol, it basically converts any effectively falsey value into an empty array. I think it's not a good practice. Let's imagine someone by mistake passes `outs=False`. Should we support this or fail? Maybe in this specific case even `outs or []` semantics is what we need, but I would double check.